File: xudongdong/QuickSort.java
import java.util.Arrays;

class QuickSort{
    public static void main(String[] args) {
        int[] arr = {11, 2, 43, 44, 25, 16, 97, 8, 19};
        System.out.println(Arrays.toString(arr));
        quick_sort(arr, 0, arr.length - 1);
        System.out.println(Arrays.toString(arr));

    }

    public static void quick_sort(int arr[], int l, int r) {
        if (l >= r) {
            return;
        }

        int i = l;
        int j = r;
        int key = arr[l];

        while (i < j) {
            while (i < j && arr[j] >= key) {
                j--;
            }

            if (i < j) {
                arr[i] = arr[j];
                i++;
            }

            while (i < j && arr[i] < key) {
                i++;
            }

            if (i < j) {
                arr[j] = arr[i];
                j--;
            }

            arr[i] = key;
            quick_sort(arr, l, i - 1);
            quick_sort(arr, i + 1, r);
        }
    }
}

File: cuican/sort/quickSort.js
const exchange = require('./util')

const sort = (array, lo, hi) => {
    let i = lo
    for (let j = lo; j <= hi; j++) {
        if (array[j] < array[hi]) {
            exchange(array, i, j)
            i++
        }
    }
    exchange(array, i, hi)
    return i
}
const quick = (array, lo, hi) => {
    if (lo < hi) {
        const i = sort(array, lo, hi)
        quick(array, lo, i - 1)
        quick(array, i+1, hi)
    }
}
const quickSort = (array) => {
    if (!array || array.length === 0) return
    quick(array, 0, array.length - 1)
    return array
}

console.log(quickSort([3,6,8,2,1]))
